### PR TITLE
Add remote control copy for objects (works if no objects are selected…

### PR DESCRIPTION
**Pull request by tsmock**
…, but just zooms to area)

Signed-off-by: Taylor Smock <taylor.smock@kaart.com>

----

It would be nice to have a feature that copies a remote control URL for the objects in the viewport and the viewport itself. This should make it slightly easier to point someone at a specific area with respect to an object.

### CODE REVIEW

**Comment by tsmock** [comment]
It also copies multiple urls to clipboard when multiple objects are selected.

File: src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java [hunk 1]
```diff
@@ -1,26 +1,34 @@
...
+    public static final String COPY = "Copy: {0}";
 
     public static void copyUser(String user) {
         if (!user.isEmpty()) {
             String linkUser = "https://www.openstreetmap.org/user/" + user;
             StringSelection selection = new StringSelection(linkUser);
             Clipboard clipboard = Toolkit.getDefaultToolkit().getSystemClipboard();
             clipboard.setContents(selection, selection);
-            new Notification(tr("Copy: " + linkUser)).setIcon(JOptionPane.INFORMATION_MESSAGE).setDuration(Notification.TIME_SHORT).show();
+            new Notification(tr(COPY, linkUser)).setIcon(JOptionPane.INFORMATION_MESSAGE).setDuration(Notification.TIME_SHORT).show();
```
Comment (don-vip) [inline src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java:36]: Translations do not work this way, you have to mark strings for translations using `marktr()`

File: src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java [hunk 2]
```diff
@@ -45,11 +53,11 @@ static void openinBrowserUserOsmComments(String user) {
 
     public static void copyChangeset(String idChangeset) {
         if (!idChangeset.isEmpty()) {
-            String linkchangeset = "https://www.openstreetmap.org/changeset/" + idChangeset;
+            String linkchangeset = "https://www.openstreetmap.org/changeset/".concat(idChangeset);
```
Comment (don-vip) [inline src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java]: why this change?

File: src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java [hunk 3]
```diff
@@ -64,26 +72,75 @@ public static void openinBrowserChangesetMap(String idChangeset) {
...
+        StringSelection selection = new StringSelection(linkobjid);
+        Clipboard clipboard = Toolkit.getDefaultToolkit().getSystemClipboard();
+        clipboard.setContents(selection, selection);
+        new Notification(tr(COPY, linkobjid)).setIcon(JOptionPane.INFORMATION_MESSAGE).setDuration(Notification.TIME_SHORT).show();
+    }
+    
+    public static void copyRemoteControl(List<AllOsmObjInfo> osmObjInfo) {
+    	if (osmObjInfo == null || osmObjInfo.isEmpty()) return;
+    	StringBuilder builder = new StringBuilder();
+        builder.append("http://127.0.0.1:8111/load_and_zoom?");
```
Comment (don-vip) [inline src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java]: instead of hardcoding it, it should be retrieved from JOSM core

File: src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java [hunk 4]
```diff
@@ -64,26 +72,75 @@ public static void openinBrowserChangesetMap(String idChangeset) {
...
+        clipboard.setContents(selection, selection);
+        new Notification(tr(COPY, remoteControl)).setIcon(JOptionPane.INFORMATION_MESSAGE).setDuration(Notification.TIME_SHORT).show();
+
     }
 
-    public static void openinBrowserIdobj(String typeObj, String idobj) {
-        if (typeObj != null && !idobj.isEmpty()) {
-            openInBrowser("https://www.openstreetmap.org/" + typeObj + "/" + idobj);
+    public static void openinBrowserIdobj(List<AllOsmObjInfo> osmObjInfo) {
+        if (osmObjInfo.size() > 5 || osmObjInfo.isEmpty()) return;
```
Comment (don-vip) [inline src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java]: why 5 ?

**Review by don-vip — CHANGES_REQUESTED** [review]

File: src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java (hunk 2 above)
Comment (tsmock) [reply src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java]: I think one of my lint tools was complaining about it (Sonarlint maybe?)
I can change it back, if needed.

File: src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java (hunk 3 above)
Comment (tsmock) [reply src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java]: Fair enough. I'll make the change.

File: src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java (hunk 4 above)
Comment (tsmock) [reply src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java]: I picked it arbitrarily. I should probably change it to be a config option.

File: src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java (hunk 1 above)
Comment (tsmock) [reply src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java:36]: I'll fix this.

**Comment by tsmock** [comment]
Ping

**Comment by don-vip** [comment]
@tsmock there are now conflicts can you please resolve them?

**Comment by tsmock** [comment]
The conflicts should be resolved.

Note: I went ahead and updated the osmcha link (`osmcha.mapbox.com` -> `osmcha.org`)

File: src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java [hunk 5]
```diff
@@ -39,17 +52,17 @@ public static void openinBrowserUserNeis(String user) {
...
         if (!user.isEmpty()) {
-            openInBrowser( "https://www.mapbox.com/osm-comments/#/changesets/?q=users:".concat(user));
+            openInBrowser("https://www.mapbox.com/osm-comments/#/changesets/?q=users:".concat(user));
         }
     }
 
     public static void copyChangeset(String idChangeset) {
         if (!idChangeset.isEmpty()) {
-            String linkchangeset = "https://www.openstreetmap.org/changeset/" + idChangeset;
+            String linkchangeset = "https://www.openstreetmap.org/changeset/".concat(idChangeset);
```
Comment (don-vip) [inline src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java]: Config.getUrls().getBaseBrowseUrl() should be used instead of hardcoding https://www.openstreetmap.org

File: src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java [hunk 6]
```diff
@@ -61,29 +74,93 @@ public static void openinBrowserChangeset(String idChangeset) {
...
-            StringSelection selection = new StringSelection(linkobjid);
-            Clipboard clipboard = Toolkit.getDefaultToolkit().getSystemClipboard();
-            clipboard.setContents(selection, selection);
-            new Notification(tr("Copy: " + linkobjid)).setIcon(JOptionPane.INFORMATION_MESSAGE).setDuration(Notification.TIME_SHORT).show();
+    public static void copyIdobj(List<AllOsmObjInfo> osmObjInfo) {
+        if (osmObjInfo == null || osmObjInfo.isEmpty()) return;
+        String linkobjid = "";
+        StringBuilder builder = new StringBuilder();
+        for (AllOsmObjInfo object : osmObjInfo) {
+            builder.append("https://www.openstreetmap.org/")
```
Comment (don-vip) [inline src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java]: Config.getUrls().getBaseBrowseUrl() should be used instead of hardcoding https://www.openstreetmap.org

File: src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java [hunk 7]
```diff
@@ -61,29 +74,93 @@ public static void openinBrowserChangeset(String idChangeset) {
...
 
-    public static void openinBrowserIdobj(String typeObj, String idobj) {
-        if (typeObj != null && !idobj.isEmpty()) {
-            openInBrowser("https://www.openstreetmap.org/" + typeObj + "/" + idobj);
+    public static void openinBrowserIdobj(List<AllOsmObjInfo> osmObjInfo) {
+        if (osmObjInfo.size() > OSMObjInfoPlugin.MAXIMUM_SELECTION.get() || osmObjInfo.isEmpty())
+            return;
+        for (AllOsmObjInfo info : osmObjInfo) {
+            if (info.typeObj != null && !info.idObject.isEmpty()) {
+                openInBrowser("https://www.openstreetmap.org/" + info.typeObj + "/" + info.idObject);
```
Comment (don-vip) [inline src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java]: Config.getUrls().getBaseBrowseUrl() should be used instead of hardcoding https://www.openstreetmap.org

**Review by don-vip — CHANGES_REQUESTED** [review]

File: src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java (hunk 5 above)
Comment (tsmock) [reply src/org/openstreetmap/josm/plugins/osmobjinfo/OSMObjInfoActions.java]: Fair enough. I think I was just reusing the pre-existing code.

**Comment by tsmock** [comment]
OK. I've switched everything to the Config.getUrls(), and I also changed the URL for changeset comments. Apparently Mapbox discontinued the osm-comments service, so I just replaced it with a link for neis-one.org.
